extensions/mssql/src/objectExplorerNodeProvider/webhdfs.ts

```diff
@@ -0,0 +1,717 @@
+		if (!argValue) {
+			throw new Error(
+				util.format(
+					localize('webhdfs.undefinedArgument', '\'%s\' is undefined.'),
+					argName
+				)
+			);
+		}
+	}
+
+	/**
+	 * Generate WebHDFS REST API endpoint URL for given operation
+	 *
+	 * @param {string} operation WebHDFS operation name
+	 * @param {string} path
+	 * @param {object} params
+	 * @returns {string} WebHDFS REST API endpoint URL
+	 */
+	private getOperationEndpoint(operation: string, path: string, params?: object): string {
+		let endpoint = this._url;
+		endpoint.pathname = this._opts.path + path;
+		let searchOpts = Object.assign(
+			{ 'op': operation },
+			this._opts.user ? { 'user.name': this._opts.user } : {},
+			params || {}
+		);
+		endpoint.search = querystring.stringify(searchOpts);
+		return url.format(endpoint);
+	}
+
+	/**
+	 * Gets localized status message for given status code
+	 *
+	 * @param {number} statusCode Http status code
+	 * @returns {string} Status message
+	 */
+	private toStatusMessage(statusCode: number): string {
+		let statusMessage: string = undefined;
+		switch (statusCode) {
+			case 400: statusMessage = localize('webhdfs.httpError400', 'Bad Request'); break;
+			case 401: statusMessage = localize('webhdfs.httpError401', 'Unauthorized'); break;
+			case 403: statusMessage = localize('webhdfs.httpError403', 'Forbidden'); break;
+			case 404: statusMessage = localize('webhdfs.httpError404', 'Not Found'); break;
+			case 500: statusMessage = localize('webhdfs.httpError500', 'Internal Server Error'); break;
+			// TODO: define more messages here
+			default: break;
+		}
+		return statusMessage;
+	}
+
+	/**
+	 * Retrieve error message from response
+	 *
+	 * @param {request.Response} response Response
+	 * @param {boolean} strict If set true then RemoteException must be present in the body
+	 * @returns {string} Error message
+	 */
+	private retrieveErrorMessage(response: request.Response, strict: boolean = false): string {
+		let errorMessage: string = undefined;
+		if (!strict) {
+			errorMessage = this.toStatusMessage(response.statusCode);
+		} else {
+			let responseBody = response.body;
+			if (typeof responseBody === 'string') {
+				try {
+					responseBody = JSON.parse(responseBody);
+				} catch { }
+			}
+
+			if (responseBody && responseBody.hasOwnProperty('RemoteException')) {
+				errorMessage = responseBody.RemoteException.message;
+			}
+		}
+
+		if (!errorMessage) {
+			errorMessage = localize('webhdfs.unknownError', 'Unknown Error');
+		}
+
+		return errorMessage;
+	}
+
+	/**
+	 * Parse error state from response and return valid Error object
+	 *
+	 * @param {request.Response} response Response
+	 * @param {boolean} strict If set true then RemoteException must be present in the body
+	 * @returns {Error} Error object
+	 */
+	private parseError(response: request.Response, strict: boolean = false): Error {
+		let errorMessage: string = this.retrieveErrorMessage(response, strict);
+		return new Error(errorMessage);
```

Comment:
Can we please subclass Error and return a status code too? Upstream callers may need to convert from that to an action. It's frustrating that we lose it.

Reply:
For example: How do we know to convert from a error code to "Login Failed" which is what we need? And prompt for username / password on the Data Services node expansion failing?

Reply:
fixed